File: algorithm/cpp/102.cpp
/**
* Definition for a binary tree node.
* struct TreeNode {
*     int val;
*     TreeNode *left;
*     TreeNode *right;
*     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
* };
*/

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if (!root) return ans;
        vector<int> cr;
        queue<TreeNode *> q;
        q.push(root);
        while (!q.empty()) {
            for (int qs = q.size(); qs > 0; qs--) {
                root = q.front(); q.pop();
                cr.push_back(root->val);
                if (root->left) q.push(root->left);
                if (root->right) q.push(root->right);
            }
            ans.push_back(move(cr));
        }
        return ans;
    }
};

File: algorithm/golang/102.go

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Queue struct {
    l *list.List
}

func New() *Queue {
    return &Queue{list.New()}
}

func (q *Queue) Push(v interface{}) {
    q.l.PushBack(v)
}
func (q *Queue) Front() interface{} {
    e := q.l.Front()
    if e != nil {
        return e.Value
    }
    return nil
}

func (q *Queue) Pop() interface{} {
    e := q.l.Front()
    if e != nil {
        q.l.Remove(e)
        return e.Value
    }
    return nil
}

func (q *Queue) Len() int {
    return q.l.Len()
}
func (q *Queue) Empty() bool {
    return q.Len() == 0
}

func (q *Queue) Clear() {
    q.l.Init()
}

func levelOrder(root *TreeNode) [][]int {
    ans := make([][]int, 0)
    if root == nil {
        return ans
    }
    q := New()
    q.Push(root)
    for !q.Empty() {
        cr := make([]int, 0)
        for s := q.Len(); s > 0; s-- {
           root = q.Pop().(*TreeNode)
           cr = append(cr, root.Val)
           if root.Left != nil {
                q.Push(root.Left)
            }
            if root.Right != nil {
                q.Push(root.Right)
            }
        }
        ans = append(ans, cr)
    }

    return ans
}
